### PR TITLE
DYN-6489 Dynamo Zoom Extents behavior has changed

### PR DESCRIPTION
**Pull request by saintentropy**
### Purpose

Helix-Toolkit made an adjustment in the `ZoomExtents()` algorithm in the 2.24.0 Release.  Specifically, the change was introduced here -> https://github.com/helix-toolkit/helix-toolkit/pull/1921 and was included in the 2.24.0 release.  Previously the `ZoomExtents()` algorithm was never very accurate or predictable in that results would vary depending on the size and height/width ratio of background preview view window.  This change however makes the calculated ZoomExtents appear fully incorrect for some cases.   Subsequent to this code change, the Dev branch in helix-toolkit now has a dramatically improved algorithm for ZoomExtents.  The PR is found here -> https://github.com/helix-toolkit/helix-toolkit/pull/2003.  I assume this new algorithm will be available in the 2.25.0 release and is testable currently with the helix-toolkit daily builds.  This PR pulls the relevant implementation into Dynamo for use until the release of 2.25.0 when we can presumably call the updated methods directly.  In testing, ZoomExtents is now fully predictable regardless of background preview view window size or height/width ratio.

This PR also takes the opportunity to utilize the native helix-toolkit bounding box calculation for scene objects.  This was not available 8 years ago when we added our own implementation.  By using the native methods, we also solve the issue where instance geometry was not correctly considered for ZoomExtents calculations.

Also for some reason we have always been clipping the top 20 pixels off of the background preview 🤷.  Now you notice it so I removed that margin. 

Improved Result
image

Previous Behavior

image

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes

Zoom Extents for the background preview is now more accurate and predictable regardless of the size of the Dynamo application window.

### Reviewers

@aparajit-pratap @mjkkirschner 

### FYIs

@jnealb @twastvedt

### CODE REVIEW

**Comment by saintentropy** [comment]
Old clipping
image

vs New 
image

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 1]
```diff
@@ -1360,7 +1360,7 @@ private void DynamoView_Loaded(object sender, EventArgs e)
             BackgroundPreview = new Watch3DView { Name = BackgroundPreviewName };
             background_grid.Children.Add(BackgroundPreview);
             BackgroundPreview.DataContext = dynamoViewModel.BackgroundPreviewViewModel;
-            BackgroundPreview.Margin = new System.Windows.Thickness(0, 20, 0, 0);
+            BackgroundPreview.Margin = new System.Windows.Thickness(0, 0, 0, 0);
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:1363]: any problem just removing this line?

File: src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs [hunk 2]
```diff
@@ -173,9 +173,14 @@ private void ViewLoadedHandler(object sender, RoutedEventArgs e)
         private void ViewModel_RequestZoomToFit(BoundingBox bounds)
         {
             var prevcamDir = watch_view.Camera.LookDirection;
-            watch_view.ZoomExtents(bounds.ToRect3D(.05));
+            var camera = watch_view.Camera as HelixToolkit.Wpf.SharpDX.PerspectiveCamera;
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:176]: it's unlikely but an extension could potentially set the camera in ortho mode, will this change the type of camera? Maybe it's just best to do a check with pattern match:

if( watch_view.Camera is HelixToolkit.Wpf.SharpDX.PerspectiveCam perspectiveCam){
}

File: src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs [hunk 3]
```diff
@@ -173,9 +173,14 @@ private void ViewLoadedHandler(object sender, RoutedEventArgs e)
...
+
+            //Todo, Call the equivalent method in Helix on adoption of next release.
+            ZoomExtents(camera, (float)(watch_view.ActualWidth / watch_view.ActualHeight), bounds, out var pos, out var look, out var up);
+            camera.AnimateTo(pos.ToPoint3D(), look.ToVector3D(), up.ToVector3D(), 0);
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:180]: hmm, if you remove this line - does the camera update?
Also, neat? Is this worth setting to some value?

File: src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs [hunk 4]
```diff
@@ -184,6 +189,117 @@ private void ViewModel_RequestZoomToFit(BoundingBox bounds)
             }
         }
 
+        #region ZoomExtents
+
+        //This implementation is found in the current dev branch of Helix-Toolkit (https://github.com/helix-toolkit/helix-toolkit/tree/develop) with the assumption it will be included in the 2.25.0 release
+        //The PR including these changes "Re-implmenet zoom extents in sharpdx versions" is found here -> https://github.com/helix-toolkit/helix-toolkit/pull/2003
+        //Specifically the code included here is located in https://github.com/holance/helix-toolkit/blob/develop/Source/HelixToolkit.SharpDX.Shared/Extensions/CameraCoreExtensions.cs
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:196]: can you add a link to a specific commit?

**Review by mjkkirschner — APPROVED** [review]
A couple questions, but it looks good otherwise, I think we should add an image comparison test if one doesn't exist already.

We likely also need a followup if one already exists to see if we can make the system more robust.

I think whoever takes the current task can add that test or file that followup task.

File: src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs (hunk 3 above)
Comment (saintentropy) [reply src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:180]: Without this line then nothing happens.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs (hunk 1 above)
Comment (saintentropy) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:1363]: Nope

File: src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs (hunk 2 above)
Comment (saintentropy) [reply src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:176]: Makes sence

**Comment by mjkkirschner** [comment]
@saintentropy I've pushed to this branch and will merge / cherry pick after the tests pass.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 5]
```diff
@@ -3006,27 +3006,23 @@ internal static Rect3D ToRect3D(this BoundingBox bounds, double minRectSize = 0.
...
         internal static BoundingBox Bounds(this GeometryModel3D geom, float defaultBoundsSize = 5.0f)
         {
-            if (geom.Geometry.Positions == null || geom.Geometry.Positions.Count == 0)
-            {
-                return new BoundingBox();
-            }
+            var bounds = geom.Bounds;
 
-            if (geom.Geometry.Positions.Count > 1)
+            if(Math.Abs(bounds.Size.LengthSquared()) < defaultBoundsSize * defaultBoundsSize * 3)
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs:3011]: A square of any number will always be greater than zero, why do we need `Math.Abs`?

Comment (aparajit-pratap) [reply src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs:3011]: Can you explain this formula, where do you get `3` from?

File: src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs [hunk 6]
```diff
@@ -173,17 +173,136 @@ private void ViewLoadedHandler(object sender, RoutedEventArgs e)
         private void ViewModel_RequestZoomToFit(BoundingBox bounds)
         {
             var prevcamDir = watch_view.Camera.LookDirection;
-            watch_view.ZoomExtents(bounds.ToRect3D(.05));
-            //if after a zoom the camera is in an undefined position or view direction, reset it.
-            if(watch_view.Camera.Position.ToVector3().IsUndefined() || 
-                watch_view.Camera.LookDirection.ToVector3().IsUndefined() || 
-                watch_view.Camera.LookDirection.Length == 0)
+            if (watch_view.Camera is HelixToolkit.Wpf.SharpDX.PerspectiveCamera perspectiveCam)
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:176]: Do we need to account for the case where the camera is Orthographic?

Comment (mjkkirschner) [reply src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:176]: hmm, let me look into what happened before.

Comment (mjkkirschner) [reply src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:176]: yeah, your're right @aparajit-pratap that would have been supported before, I'll move in the code for ortho cameras as well from helix.

Comment (mjkkirschner) [reply src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:180]: Screenshot 2023-12-12 at 2 40 15 PM
Screenshot 2023-12-12 at 2 40 12 PM

done.

Comment (aparajit-pratap) [reply src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:180]: Wait, how do we change the camera from perspective to orthographic? Is there a UI option?

**Review by aparajit-pratap — APPROVED** [review]

File: src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs (hunk 6 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs:180]: it can be done from an extension - theres a property on the helix viewport object

**Pull request merged**